### ISSUE TITLE
[gateway] [oidc] unknown (use) is currently not supported.

### ISSUE THREAD

**Issue by gianx89** [gravitee-io/issues#6184]
Hi, I'm running the latest gravitee version 3.11.

I'm tryng to integrate AM with Geoserver using its community plugin for oauth and openid.

All is correctly configured, but authentication fails (Geoserver log):

```
2021-09-16 19:32:14,355 ERROR [org.geoserver.security] - Could not Authorize OAuth2 Resource due to the following exception:
error="server_error", error_description="unknown (use) is currently not supported."
        at org.springframework.security.oauth2.provider.token.store.jwk.JwkSetConverter.createRsaJwkDefinition(JwkSetConverter.java:136)
        at org.springframework.security.oauth2.provider.token.store.jwk.JwkSetConverter.convert(JwkSetConverter.java:98)
        at org.springframework.security.oauth2.provider.token.store.jwk.JwkDefinitionSource.loadJwkDefinitions(JwkDefinitionSource.java:131)
        at org.springframework.security.oauth2.provider.token.store.jwk.JwkDefinitionSource.getDefinitionLoadIfNecessary(JwkDefinitionSource.java:93)
        at org.springframework.security.oauth2.provider.token.store.jwk.JwkVerifyingJwtAccessTokenConverter.decode(JwkVerifyingJwtAccessTokenConverter.java:103)
        at org.springframework.security.oauth2.provider.token.store.JwtTokenStore.readAuthentication(JwtTokenStore.java:71)
        at org.springframework.security.oauth2.provider.token.store.jwk.JwkTokenStore.readAuthentication(JwkTokenStore.java:123)
        at org.geoserver.security.oauth2.ValidatingOAuth2RestTemplate.validate(ValidatingOAuth2RestTemplate.java:40)
        at org.geoserver.security.oauth2.ValidatingOAuth2RestTemplate.getAccessToken(ValidatingOAuth2RestTemplate.java:30)
        at org.springframework.security.oauth2.client.OAuth2RestTemplate$$FastClassBySpringCGLIB$$ca6dc720.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:752)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.doProceed(DelegatingIntroductionInterceptor.java:136)
        at org.springframework.aop.support.DelegatingIntroductionInterceptor.invoke(DelegatingIntroductionInterceptor.java:124)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:691)
        at org.springframework.security.oauth2.client.OAuth2RestTemplate$$EnhancerBySpringCGLIB$$99393dec.getAccessToken(<generated>)
        at org.springframework.security.oauth2.client.filter.OAuth2ClientAuthenticationProcessingFilter.attemptAuthentication(OAuth2ClientAuthenticationProcessingFilter.java:105)
        at org.geoserver.security.oauth2.GeoServerOAuthAuthenticationFilter.getPreAuthenticatedPrincipal(GeoServerOAuthAuthenticationFilter.java:369)
        at org.geoserver.security.oauth2.GeoServerOAuthAuthenticationFilter.doAuthenticate(GeoServerOAuthAuthenticationFilter.java:278)
        at org.geoserver.security.oauth2.GeoServerOAuthAuthenticationFilter.doFilter(GeoServerOAuthAuthenticationFilter.java:149)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
        at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:70)
        at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:105)
        at org.geoserver.security.filter.GeoServerSecurityContextPersistenceFilter$1.doFilter(GeoServerSecurityContextPersistenceFilter.java:52)
        at org.geoserver.security.filter.GeoServerCompositeFilter$NestedFilterChain.doFilter(GeoServerCompositeFilter.java:74)
        at org.geoserver.security.filter.GeoServerCompositeFilter.doFilter(GeoServerCompositeFilter.java:91)
        at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:334)
        at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:215)
        at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:178)
        at org.geoserver.security.GeoServerSecurityFilterChainProxy.doFilter(GeoServerSecurityFilterChainProxy.java:142)
        at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)
        at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.geoserver.filters.LoggingFilter.doFilter(LoggingFilter.java:101)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.geoserver.filters.XFrameOptionsFilter.doFilter(XFrameOptionsFilter.java:77)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.geoserver.filters.GZIPFilter.doFilter(GZIPFilter.java:47)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.geoserver.filters.SessionDebugFilter.doFilter(SessionDebugFilter.java:46)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.geoserver.filters.FlushSafeFilter.doFilter(FlushSafeFilter.java:42)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:374)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:888)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1597)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.base/java.lang.Thread.run(Thread.java:834)
```

I think it's related to:

[https://github.com/spring-projects/spring-security-oauth/pull/1348](https://github.com/spring-projects/spring-security-oauth/pull/1348)

I suppose that the geoserver plugin uses spring-security-oauth: it searches for **the "use" field** ([https://datatracker.ietf.org/doc/html/rfc7517#page-6](https://datatracker.ietf.org/doc/html/rfc7517#page-6)), it doesn't find it and it crashes.

Spring guys are not willing to make it optional.

Would be it possible to add this field to **/oidc/.well-known/jwks.json** (I'm not an expert, but it would be the right place, right?).

```
{
  "keys" : [ {
    "kty" : "RSA",
    "alg" : "RS256",
    "kid" : "default",
    "x5c" : [ "" ],
    "x5t#S256" : "",
    "e" : "",
    "n" : ""
  } ]
}
```

to

```
{
  "keys" : [ {
    "kty" : "RSA",
    "alg" : "RS256",
    "kid" : "default",
    "x5c" : [ "" ],
    "x5t#S256" : "",
    "e" : "",
    "n" : "",
    "use":"sig"
  } ]
}
```

I don't know the codebase, but  think it would be pretty easy. I know the parameter is optional, but we could achieve broader compatibility.

Is there a setting to turn this feature on?

We're blocked, can you help me? @brasseld @tcompiegne

**Comment by tcompiegne** [comment]
Hi @gianx89 ,

Always happy to help, we will check to add this ASAP.

**Comment by gianx89** [comment]
Thank you so much for your reply, even at this hour. I think it will be a pretty easy "fix" that will help a lot of people.

I'm waiting for more details in the next days.

Thank you so much again.

**Comment by tcompiegne** [comment]
@gianx89 the 3.10.4 is now available.

**Comment by gianx89** [comment]
> @gianx89 the 3.10.4 is now available.

I'm 3.11, is the changed merged and released on 3.11.x as well?
